core/llm_google.go

```diff
@@ -164,91 +122,193 @@ func (c *GenaiClient) SendQuery(ctx context.Context, history []ModelMessage, too
 			genaiHistory = append(genaiHistory, content)
 		}
 	}
-	if len(model.SystemInstruction.Parts) == 0 {
-		model.SystemInstruction = nil
-	}
 
-	// Pop last message from history for SendMessage
-	if len(genaiHistory) == 0 {
-		return nil, fmt.Errorf("no user prompt")
+	if len(systemInstruction.Parts) == 0 {
+		systemInstruction = nil
 	}
 
-	userMessage := genaiHistory[len(genaiHistory)-1]
-	genaiHistory = genaiHistory[:len(genaiHistory)-1]
+	return genaiHistory, systemInstruction, nil
+}
 
-	chat := model.StartChat()
-	chat.History = genaiHistory
+// Define callbacks passed to processStreamResponse for handling stream events
+// We isolate the core stream iteration logic from the application logic (logging, metrics, final accumulation)
+type textPartHandler func(ctx context.Context, text string) error
+type tokenUsageHandler func(ctx context.Context, usageMeta *genai.UsageMetadata, usageSummary *LLMTokenUsage) error
```

Comment:
usageSummary is passed as an input, but it's actually an output.
```suggestion
type tokenUsageHandler func(ctx context.Context, usageMeta *genai.UsageMetadata) (usageSummary LLMTokenUsage, error)
```

Reply:
Might as well remove the type declaration and just inline the type in `processStreamResponse`'s signature.